### PR TITLE
[tlse] update TLS/CA config

### PR DESCRIPTION
**Pull request by stuggi**
changes top level tls config section to reflect ingress and podlevel tls termination and allows to customize the duration and renewBefore for each of the CAs and Certs issued for a CA.

Also:
* creates CA for OVN
* fixes an issue where service cert was not created when tls on ingress
  was disabled.
*  Allow using custom tlsConfig.API.Public.SecretName when no ingress used
   When also using LoadBalancer (MetalLB) for public endpoints this change allows to use the service configs tls.API.Public.SecretName to reference a secret holding a custom TLS cert. The secret must contain at least tls.key and tls.crt. The custom CA should be added to the bundle using the secret reference in the osctlplane crd.

Depends-On: https://github.com/openstack-k8s-operators/lib-common/pull/471

Jira: [OSPRH-5342](https://issues.redhat.com//browse/OSPRH-5342)

### CODE REVIEW

**Comment by stuggi** [comment]
the tls custom config would now look like this on a ctlplane CR. duration, renewBefore are defaulted.

```
apiVersion: core.openstack.org/v1beta1
kind: OpenStackControlPlane
metadata:
  name: openstack-network-isolation-ceph
spec:
  tls:
    ingress:
      ca:
        duration: 43800h
      cert:
        duration: 1h
        renewBefore: 55m
      enabled: true
    podLevel:
      default:
        ca:
          duration: 43800h
        cert:
          duration: 300h
          renewBefore: 5m
      enabled: true
      ovn:
        ca:
          duration: 43800h
        cert:
          duration: 8760h
```

File: config/samples/core_v1beta1_openstackcontrolplane_network_isolation_tlse.yaml [hunk 1]
```diff
@@ -6,11 +6,10 @@ spec:
   secret: osp-secret
   storageClass: local-storage
   tls:
-    endpoint:
-      internal:
-        enabled: true
-      public:
-        enabled: true
+    ingress:
```
Comment (olliewalsh) [inline config/samples/core_v1beta1_openstackcontrolplane_network_isolation_tlse.yaml:9]: nit: don't need to explicitly set ingress to true

File: apis/core/v1beta1/openstackcontrolplane_types.go [hunk 2]
```diff
@@ -200,12 +203,67 @@ type TLSSection struct {
...
 	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:booleanSwitch"}
 	// Enabled - Whether TLS should be enabled for endpoint type
 	Enabled bool `json:"enabled"`
+
+	// +kubebuilder:validation:optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec
+	CertSection `json:",inline"`
+}
+
+// TLSPodLevelConfig defines the desired state of the TLS configuration for TLS termination at the pod level
```
Comment (olliewalsh) [inline apis/core/v1beta1/openstackcontrolplane_types.go:218]: nit: technically libvirt isn't a pod.... but can't think of a better name

File: pkg/openstack/ca.go [hunk 3]
```diff
@@ -94,49 +94,60 @@ func ReconcileCAs(ctx context.Context, instance *corev1.OpenStackControlPlane, h
...
+	}
 
-		instance.Status.TLS.CAList = append(instance.Status.TLS.CAList, status)
+	// create CA for ovn
+	ctrlResult, err = ensureRootCA(
+		ctx,
+		instance,
+		helper,
+		issuerReq,
+		tls.DefaultCAPrefix+"ovn",
```
Comment (olliewalsh) [inline pkg/openstack/ca.go:139]: label uses "ovndb" so maybe use this for the CA name to keep it consistent

File: pkg/openstack/ca.go [hunk 4]
```diff
@@ -94,49 +94,60 @@ func ReconcileCAs(ctx context.Context, instance *corev1.OpenStackControlPlane, h
...
+		map[string]string{certmanager.RootCAIssuerOvnDBLabel: ""},
+		bundle,
+		caOnlyBundle,
+		instance.Spec.TLS.PodLevel.Ovn.Ca,
```
Comment (olliewalsh) [inline pkg/openstack/ca.go:143]: label users "ovndb" so maybe use OvnDb to keep is consistent

File: apis/bases/core.openstack.org_openstackcontrolplanes.yaml [hunk 5]
```diff
@@ -16499,20 +16499,84 @@ spec:
...
+                            type: string
+                          renewBefore:
+                            type: string
+                        type: object
+                      enabled:
+                        type: boolean
+                    type: object
+                  podLevel:
+                    properties:
+                      default:
```
Comment (olliewalsh) [inline apis/bases/core.openstack.org_openstackcontrolplanes.yaml:16544]: kindof implies that ovn will default to this too but IIUC it will not

**Comment by stuggi** [comment]
recheck

File: apis/bases/core.openstack.org_openstackcontrolplanes.yaml (hunk 5 above)
Comment (stuggi) [reply apis/bases/core.openstack.org_openstackcontrolplanes.yaml]: as discussed, renamed to internal

File: pkg/openstack/ca.go (hunk 3 above)
Comment (stuggi) [reply pkg/openstack/ca.go:139]: https://github.com/openstack-k8s-operators/lib-common/pull/472

**Comment by stuggi** [comment]
rebased

**Comment by stuggi** [comment]
recheck

**Comment by stuggi** [comment]
recheck

**Comment by stuggi** [comment]
nova is still in deploy progress. both cell conductors do not have db access

cell1 conductor can not access the db https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/nova-cell1-conductor-0/logs/nova-cell1-conductor-conductor.log

```
2024-03-07 12:45:12.803 1 ERROR nova.context sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1045, "Access denied for user 'nova_cell1'@'192.168.122.100' (using password: YES)")
```

the db create and account create job for cell1 seems has finished ok 
https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/nova-cell1-db-create-q8qfv/nova-cell1-db-create-q8qfv-describe

Same for conductor in cell0 https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/nova-cell0-conductor-0/logs/nova-cell0-conductor-conductor.log

nova-scheduler and nova-api seem to be fine

**Comment by stuggi** [comment]
Galera pod logs show abort erros for both instances

https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/openstack-galera-0/logs/galera.log

https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/openstack-cell1-galera-0/logs/galera.log

```
2024-03-07 12:39:12 120 [Warning] Aborted connection 120 to db: 'keystone' user: 'keystone' host: '10.217.1.15' (Got an error reading communication packets)
2024-03-07 12:39:20 121 [Warning] Aborted connection 121 to db: 'keystone' user: 'keystone' host: '10.217.1.16' (Got an error reading communication packets)
2024-03-07 12:39:27 124 [Warning] Aborted connection 124 to db: 'keystone' user: 'keystone' host: '10.217.1.18' (Got an error reading communication packets)
2024-03-07 12:40:35 131 [Warning] Aborted connection 131 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
2024-03-07 12:40:35 130 [Warning] Aborted connection 130 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
2024-03-07 12:40:35 129 [Warning] Aborted connection 129 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
2024-03-07 12:40:35 134 [Warning] Aborted connection 134 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
2024-03-07 12:40:53 152 [Warning] Aborted connection 152 to db: 'barbican' user: 'barbican' host: '10.217.1.25' (Got an error reading communication packets)
2024-03-07 12:42:57 202 [Warning] Aborted connection 202 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
2024-03-07 12:42:57 203 [Warning] Aborted connection 203 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
2024-03-07 12:42:58 204 [Warning] Aborted connection 204 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
2024-03-07 12:43:49 218 [Warning] Aborted connection 218 to db: 'placement' user: 'placement' host: '10.217.1.44' (Got an error reading communication packets)
2024-03-07 12:44:40 236 [Warning] Aborted connection 236 to db: 'nova_api' user: 'nova_api' host: '10.217.1.53' (Got an error reading communication packets)
2024-03-07 12:44:42 239 [Warning] Aborted connection 239 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.53' (Got an error reading communication packets)
2024-03-07 12:45:03 247 [Warning] Aborted connection 247 to db: 'nova_api' user: 'nova_api' host: '10.217.1.56' (Got an error reading communication packets)
2024-03-07 12:45:05 249 [Warning] Aborted connection 249 to db: 'nova_api' user: 'nova_api' host: '10.217.1.56' (Got an error reading communication packets)
2024-03-07 12:45:11 248 [Warning] Aborted connection 248 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.61' (Got an error reading communication packets)
2024-03-07 12:45:11 246 [Warning] Aborted connection 246 to db: 'nova_api' user: 'nova_api' host: '10.217.1.61' (Got an error reading communication packets)
2024-03-07 12:45:24 271 [Warning] Aborted connection 271 to db: 'nova_api' user: 'nova_api' host: '10.217.1.68' (Got an error reading communication packets)
2024-03-07 12:45:26 272 [Warning] Aborted connection 272 to db: 'nova_api' user: 'nova_api' host: '10.217.1.68' (Got an error reading communication packets)
2024-03-07 12:45:32 270 [Warning] Aborted connection 270 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 265 [Warning] Aborted connection 265 to db: 'nova_api' user: 'nova_api' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 267 [Warning] Aborted connection 267 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 269 [Warning] Aborted connection 269 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 268 [Warning] Aborted connection 268 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 266 [Warning] Aborted connection 266 to db: 'nova_api' user: 'nova_api' host: '10.217.1.65' (Got an error reading communication packets)
2024-03-07 12:45:32 261 [Warning] Aborted connection 261 to db: 'nova_api' user: 'nova_api' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 262 [Warning] Aborted connection 262 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 263 [Warning] Aborted connection 263 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 264 [Warning] Aborted connection 264 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 260 [Warning] Aborted connection 260 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 259 [Warning] Aborted connection 259 to db: 'nova_api' user: 'nova_api' host: '10.217.1.66' (Got an error reading communication packets)
2024-03-07 12:45:32 255 [Warning] Aborted connection 255 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.67' (Got an error reading communication packets)
2024-03-07 12:45:32 256 [Warning] Aborted connection 256 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.67' (Got an error reading communication packets)
2024-03-07 12:45:32 254 [Warning] Aborted connection 254 to db: 'nova_api' user: 'nova_api' host: '10.217.1.67' (Got an error reading communication packets)
2024-03-07 12:45:43 278 [Warning] Aborted connection 278 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.71' (Got an error reading communication packets)
```


cell1
```
2024-03-07 12:45:04 174 [Warning] Aborted connection 174 to db: 'nova_cell1' user: 'nova_cell1' host: '10.217.1.62' (Got an error reading communication packets)
```

**Comment by stuggi** [comment]
> Galera pod logs show abort erros for both instances
> 
> https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/openstack-galera-0/logs/galera.log
> 
> https://logserver.rdoproject.org/93/693/457c06271939e17c7af54e1a3bf25aaed98c6604/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/44455a3/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/openstack-cell1-galera-0/logs/galera.log
> 
> ```
> 2024-03-07 12:39:12 120 [Warning] Aborted connection 120 to db: 'keystone' user: 'keystone' host: '10.217.1.15' (Got an error reading communication packets)
> 2024-03-07 12:39:20 121 [Warning] Aborted connection 121 to db: 'keystone' user: 'keystone' host: '10.217.1.16' (Got an error reading communication packets)
> 2024-03-07 12:39:27 124 [Warning] Aborted connection 124 to db: 'keystone' user: 'keystone' host: '10.217.1.18' (Got an error reading communication packets)
> 2024-03-07 12:40:35 131 [Warning] Aborted connection 131 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
> 2024-03-07 12:40:35 130 [Warning] Aborted connection 130 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
> 2024-03-07 12:40:35 129 [Warning] Aborted connection 129 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
> 2024-03-07 12:40:35 134 [Warning] Aborted connection 134 to db: 'keystone' user: 'keystone' host: '10.217.1.19' (Got an error reading communication packets)
> 2024-03-07 12:40:53 152 [Warning] Aborted connection 152 to db: 'barbican' user: 'barbican' host: '10.217.1.25' (Got an error reading communication packets)
> 2024-03-07 12:42:57 202 [Warning] Aborted connection 202 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
> 2024-03-07 12:42:57 203 [Warning] Aborted connection 203 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
> 2024-03-07 12:42:58 204 [Warning] Aborted connection 204 to db: 'glance' user: 'glance' host: '10.217.1.38' (Got an error reading communication packets)
> 2024-03-07 12:43:49 218 [Warning] Aborted connection 218 to db: 'placement' user: 'placement' host: '10.217.1.44' (Got an error reading communication packets)
> 2024-03-07 12:44:40 236 [Warning] Aborted connection 236 to db: 'nova_api' user: 'nova_api' host: '10.217.1.53' (Got an error reading communication packets)
> 2024-03-07 12:44:42 239 [Warning] Aborted connection 239 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.53' (Got an error reading communication packets)
> 2024-03-07 12:45:03 247 [Warning] Aborted connection 247 to db: 'nova_api' user: 'nova_api' host: '10.217.1.56' (Got an error reading communication packets)
> 2024-03-07 12:45:05 249 [Warning] Aborted connection 249 to db: 'nova_api' user: 'nova_api' host: '10.217.1.56' (Got an error reading communication packets)
> 2024-03-07 12:45:11 248 [Warning] Aborted connection 248 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.61' (Got an error reading communication packets)
> 2024-03-07 12:45:11 246 [Warning] Aborted connection 246 to db: 'nova_api' user: 'nova_api' host: '10.217.1.61' (Got an error reading communication packets)
> 2024-03-07 12:45:24 271 [Warning] Aborted connection 271 to db: 'nova_api' user: 'nova_api' host: '10.217.1.68' (Got an error reading communication packets)
> 2024-03-07 12:45:26 272 [Warning] Aborted connection 272 to db: 'nova_api' user: 'nova_api' host: '10.217.1.68' (Got an error reading communication packets)
> 2024-03-07 12:45:32 270 [Warning] Aborted connection 270 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 265 [Warning] Aborted connection 265 to db: 'nova_api' user: 'nova_api' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 267 [Warning] Aborted connection 267 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 269 [Warning] Aborted connection 269 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 268 [Warning] Aborted connection 268 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 266 [Warning] Aborted connection 266 to db: 'nova_api' user: 'nova_api' host: '10.217.1.65' (Got an error reading communication packets)
> 2024-03-07 12:45:32 261 [Warning] Aborted connection 261 to db: 'nova_api' user: 'nova_api' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 262 [Warning] Aborted connection 262 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 263 [Warning] Aborted connection 263 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 264 [Warning] Aborted connection 264 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 260 [Warning] Aborted connection 260 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 259 [Warning] Aborted connection 259 to db: 'nova_api' user: 'nova_api' host: '10.217.1.66' (Got an error reading communication packets)
> 2024-03-07 12:45:32 255 [Warning] Aborted connection 255 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.67' (Got an error reading communication packets)
> 2024-03-07 12:45:32 256 [Warning] Aborted connection 256 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.67' (Got an error reading communication packets)
> 2024-03-07 12:45:32 254 [Warning] Aborted connection 254 to db: 'nova_api' user: 'nova_api' host: '10.217.1.67' (Got an error reading communication packets)
> 2024-03-07 12:45:43 278 [Warning] Aborted connection 278 to db: 'nova_cell0' user: 'nova_cell0' host: '10.217.1.71' (Got an error reading communication packets)
> ```
> 
> cell1
> 
> ```
> 2024-03-07 12:45:04 174 [Warning] Aborted connection 174 to db: 'nova_cell1' user: 'nova_cell1' host: '10.217.1.62' (Got an error reading communication packets)
> ```

they are probably not related we also see aborts in a successful run as in https://logserver.rdoproject.org/90/690/b0d0bda02ab5ae4f8fb39744e8cd606d829ffa04/github-check/cifmw-data-plane-adoption-osp-17-to-extracted-crc/e50ed17/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/openstack-cell1-galera-0/logs/galera.log

**Comment by stuggi** [comment]
try bumping nova and mariadb operator as the issue is not seen in https://github.com/openstack-k8s-operators/openstack-operator/pull/690 , but kuttl might fail as in that one

**Review by olliewalsh — APPROVED** [review]
/lgtm pushed a rebase

**Pull request merged**